### PR TITLE
feat: folding for tosa.pad

### PR DESCRIPTION
**Pull request by ttjost**

### CODE REVIEW

**Review by flemairen6 — APPROVED** [review]

File: mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp [hunk 1]
```diff
@@ -1266,6 +1266,121 @@ struct TosaFoldConstantMaximum
...
+    case 8:
+      return padType<int8_t>(inputType, inputValues, paddings, padConstValue,
+                             outputType, 0);
+    case 16:
+      return padType<int16_t>(inputType, inputValues, paddings, padConstValue,
+                              outputType, 0);
+    case 32:
+      return padType<int32_t>(inputType, inputValues, paddings, padConstValue,
+                              outputType, 0);
+    case 64:
```
Comment (cmcgirr-amd) [inline mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp:1326]: The PAD spec states only up to signed 32. Do we need it then?

File: mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp [hunk 2]
```diff
@@ -1266,6 +1266,121 @@ struct TosaFoldConstantMaximum
...
+DenseElementsAttr padType(ShapedType inputType, ElementsAttr inputValues,
+                          DenseElementsAttr paddings,
+                          std::optional<DenseElementsAttr> padConstValue,
+                          ShapedType outputType, BaseType zero) {
+  BaseType padConst(zero);
+  if (padConstValue.has_value())
+    padConst = padConstValue.value().getSplatValue<BaseType>();
+
+  auto values = inputValues.getValues<BaseType>();
+  auto paddingVals = paddings.getValues<int64_t>();
```
Comment (cmcgirr-amd) [inline mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp:1279]: Interesting, the spec says int32 but the op definition in TableGen says int32 or int64 :D

File: mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp [hunk 3]
```diff
@@ -1266,6 +1266,121 @@ struct TosaFoldConstantMaximum
...
+                            outputType,
+                            APInt(baseType.getIntOrFloatBitWidth(), 0));
+    }
+  }
+
+  assert(isa<FloatType>(baseType) && "Unknown element type.");
+  FloatType fpType = cast<FloatType>(baseType);
+
+  APFloat zero(fpType.getFloatSemantics(), APInt::getZero(fpType.getWidth()));
+  return padType<APFloat>(inputType, inputValues, paddings, padConstValue,
```
[the remark this answers is not in the document]
Comment (ttjost) [reply mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp:1340]: I will update it to only support them.

File: mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp (hunk 1 above)
Comment (ttjost) [reply mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp]: Simplified to only use `APInt` and `APFloat`. Verifiers should be the ones checking for types.

File: mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp (hunk 2 above)
Comment (ttjost) [reply mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp:1279]: True!

**Review by cmcgirr-amd — APPROVED** [review]

**Pull request merged**